### ISSUE TITLE
In-context external link doesn’t work behind a proxy

### ISSUE THREAD

**Issue by crism** [anvc/scalar#100]
When a link leaves a Scalar book, and isn’t marked for a new context, Javascript intercepts it and constructs a link to the `external` handler. That handler [does some security checking](https://github.com/anvc/scalar/blob/master/system/application/controllers/book.php#L401), including that the `prev` parameter matches the Scalar server:
```php
if (!stristr($this->data['prev'], base_url())) $this->kickout();
```
However, if the Scalar server is running behind a proxy, `prev` doesn’t match the `base_url()`. Suggestions (in order of our preference):

 1. Drop the `external` handler altogether. This kind of framing context feels a little outdated and presents possible accessibility issues. External links could just supplant the Scalar context, and users can use their Back button to return.
 1. Truncate the `prev` parameter—make it a relative URL, within the scope of the Scalar book, guaranteed always to be safe.
 1. Fix this for proxy detection. Since the `prev` link is constructed browser-side, it’ll reflect the _apparent_ address of the server, not the actual address. The request parameters should provide enough information to see if the previous link is sane, or the referer [sic] maybe could be used instead of the provided `prev` parameter. (I need to investigate this option more, if desirable.)
 1. Relax the check somehow. I’m not sure how far, but it could look for same domain name instead of server name, plus URL… this would be a bit messy and provide half-baked security, hence last on this list.

In addition, the `kickout()` ought to provide a little more information to the user; as it stands, if something goes wrong, they end up at the Scalar home page, lost and confused.

**Comment by craigdietrich** [comment]
Yep!  The external page while well intention raises a number of problems.

There's two quick fixes in local_settings.php, the second field below may or may not be in yours depending on how old your Scalar install is, but you can just add it:

```
// Redlisted domains that won't let themselves be included in external.php's iframe
$config['iframe_redlist'] = array('youtube.com', 'vimeo.com', 'google.com', 'microsoft.com', 'metmuseum.org', 'bloomberg.com', 'nytimes.com', 'nationalreview.com', 'theguardian.com', 'twitter.com'); 

// If true, external hyperlinks will go directly to the external page rather than attempt to be rendered in Scalar's external.php
$config['external_direct_hyperlink'] = false;
```

Hopefully helpful!

**Comment by crism** [comment]
Thanks, @craigdietrich! The config fix works very nicely; I had overlooked that setting.

I should also add that it doesn’t work in mixed HTTP/HTTPS cases—_e.g._, the Scalar book is HTTPS but the external link is HTTP. Many browsers will block that kind of mixed content.

Seems like this is worth looking into still, but a lot less urgent for us. Thanks again.
